Sources/NIOWebSocket/WebSocketFrameDecoder.swift

```diff
@@ -0,0 +1,321 @@
+}
+
+public extension ByteBuffer {
+    /// Applies the WebSocket unmasking operation.
+    ///
+    /// - parameters:
+    ///     - maskingKey: The masking key.
+    public mutating func unmask(_ maskingKey: WebSocketMaskingKey, indexOffset: Int = 0) {
+        /// Shhhh: secretly unmasking and masking are the same operation!
+        mask(maskingKey, indexOffset: indexOffset)
+    }
+
+    /// Applies the websocket masking operation.
+    ///
+    /// - parameters:
+    ///     - maskingKey: The masking key.
+    ///     - indexOffset: An integer offset to apply to the index into the masking key.
+    ///         This is used when masking multiple "contiguous" byte buffers, to ensure that
+    ///         the masking key is applied uniformly to the collection rather than from the
+    ///         start each time.
+    public mutating func mask(_ maskingKey: WebSocketMaskingKey, indexOffset: Int = 0) {
+        self.withUnsafeMutableReadableBytes {
+            for (index, byte) in $0.enumerated() {
+                $0[index] = byte ^ maskingKey[(index + indexOffset) % 4]
+            }
+        }
+    }
+}
+
+/// The current state of the frame decoder.
+enum DecoderState {
+    /// Waiting for a frame.
+    case idle
+
+    /// The initial frame byte has been received, but the length byte
+    /// has not.
+    case firstByteReceived
+
+    /// The length byte indicates that we need to wait for the length word, and we're
+    /// currently waiting for it.
+    case waitingForLengthWord
+
+    /// The length byte indicates that we need to wait for the length qword, and
+    /// we're currently waiting for it.
+    case waitingForLengthQWord
+
+    /// The mask bit indicates we are expecting a mask key.
+    case waitingForMask
+
+    /// All the header data is complete, we are waiting for the application data.
+    case waitingForData
+}
+
+enum ParseResult {
+    case insufficientData
+    case continueParsing
+    case result(WebSocketFrame)
+}
+
+/// An incremental websocket frame parser.
+///
+/// This parser attempts to parse a websocket frame incrementally, keeping as much parsing state around as possible to ensure that
+/// we don't repeatedly partially parse the data.
+struct WSParser {
+    var firstByte: UInt8? = nil
+    var length: Int? = nil
+    var masked: Bool = false
+    var maskingKey: WebSocketMaskingKey? = nil
+
+    /// The current state of the decoder during incremental parse.
+    var state: DecoderState = .idle
+
+    private mutating func reset() {
+        self.state = .idle
+        self.firstByte = nil
+        self.length = nil
+        self.masked = false
+        self.maskingKey = nil
+    }
+
+    mutating func parseStep(_ buffer: inout ByteBuffer) -> ParseResult {
+        switch self.state {
+        case .idle:
+            // This is a new buffer. We want to find the first octet and save it off.
+            assert(self.firstByte == nil)
+            guard let firstByte: UInt8 = buffer.readInteger() else {
```

Comment:
I think we should use the new `buffer.readInteger(as: UInt8.self)` here and below